src/java.base/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -265,7 +291,7 @@ static MethodHandle mergeArguments(MethodHandle mh, int sourceIndex, int destInd
     }
 
 
-    static MethodHandle swapArguments(MethodHandle mh, int firstArg, int secondArg) {
+    public static MethodHandle swapArguments(MethodHandle mh, int firstArg, int secondArg) {
         MethodType mtype = mh.type();
         int[] perms = new int[mtype.parameterCount()];
         MethodType swappedType = MethodType.methodType(mtype.returnType());
```

Comment:
Instead of `MethodType::appendParameterTypes(…)` (which performs an expensive lookup of a new cached `MethodType` value per call), this method should instead use a `Class<?>[]` for the arguments, which avoids that overhead inside a loop:
```java
	public static MethodHandle swapArguments(MethodHandle mh, int firstArg, int secondArg) {
		MethodType mtype = mh.type();
		int[] perms = new int[mtype.parameterCount()];
		Class<?>[] ptypes = new Class<?>[perms.length];
		for (int i = 0 ; i < perms.length ; i++) {
			int dst = i;
			if (i == firstArg) dst = secondArg;
			else if (i == secondArg) dst = firstArg;
			perms[i] = dst;
			ptypes[i] = mtype.parameterType(dst);
		}
		// This should use `JavaLangInvokeAccess` to invoke the internal
		// `MethodType.methodType(Class<?> rtype, Class<?>[] ptypes, boolean trusted)`
		// method with a `trusted` value of `true`:
		MethodType swappedType = MethodType.methodType(mtype.returnType(), ptypes);
		return permuteArguments(mh, swappedType, perms);
	}
```

Reply:
I don’t (yet) have a **JBS** account, so it’s painful to file issues because the **Java Bug Report** website[^1] doesn’t support neither **Markdown** nor **Jira** formatting (and also the `JI‑*` to `JDK‑*` issue transfer waiting time).

[^1]: https://bugreport.java.com/bugreport/

Reply:
Filed as: [JI‑9075066](https://bugs.openjdk.org/browse/JI-9075066)